File: packages/telemetry/src/errors/telemetry-adapter-error.ts
export class TelemetryAdapterError extends Error {
	constructor(
		message: string,
		readonly context?: Record<string, unknown>,
	) {
		super(message);
		this.name = 'TelemetryAdapterError';
	}
}

export class TelemetryApiError extends TelemetryAdapterError {
	constructor(
		message: string,
		readonly status: number,
		readonly body?: unknown,
		context?: Record<string, unknown>,
	) {
		super(message, context);
		this.name = 'TelemetryApiError';
	}
}

export class UnsupportedMetricTypeError extends TelemetryAdapterError {
	constructor(
		message: string,
		readonly metricType: string,
		context?: Record<string, unknown>,
	) {
		super(message, context);
		this.name = 'UnsupportedMetricTypeError';
	}
}

File: kotlin/modules/telemetry/src/main/kotlin/com/ffpromo/telemetry/errors/TelemetryAdapterError.kt
package com.ffpromo.telemetry.errors

open class TelemetryAdapterError(
    message: String,
    open val context: Map<String, Any?>? = null,
    cause: Throwable? = null,
) : Exception(message, cause)

class TelemetryApiError(
    message: String,
    val status: Int,
    val body: Any? = null,
    context: Map<String, Any?>? = null,
) : TelemetryAdapterError(message, context)

class UnsupportedMetricTypeError(
    message: String,
    val metricType: String,
    context: Map<String, Any?>? = null,
) : TelemetryAdapterError(message, context)
